src/future/value_initializer.rs

```diff
@@ -156,6 +158,40 @@ where
         self.do_try_init(&key, type_id, init, post_init).await
     }
 
+    /// # Panics
+    /// Panics if the `init` future has been panicked.
+    pub(super) async fn optionally_init_or_read<F>(
+        &self,
+        key: Arc<K>,
+        init: F,
+    ) -> InitResult<V, OptionallyNone>
+    where
+        F: Future<Output = Option<V>>,
+    {
+        let type_id = TypeId::of::<OptionallyNone>();
+
+        // This closure will be called after the init closure has returned a value.
+        // It will convert the returned value (from init) into an InitResult.
+        let post_init = |key, value: Option<V>, mut guard: WaiterGuard<'_, K, V, S>| match value {
+            Some(value) => {
+                guard.set_waiter_value(WaiterValue::Ready(Ok(value.clone())));
+                InitResult::Initialized(value)
+            }
+            None => {
+                // `value` can be either `Some` or `None`. For `None` case, without
+                // change the existing API too much, we will need to convert `None`
+                // to Arc<E> here. `Infalliable` could not be instantiated. So it
+                // might be good to use an empty struct to indicate the error type.
+                let err: ErrorObject = Arc::new(OptionallyNone);
```

Comment:
> `Infalliable` could not be instantiated.

Thanks. You are right. `Infallible` is a unit enum and not struct, so it could not be instantiated.